### PR TITLE
Faster sync tests

### PR DESCRIPTION
**Pull request by asdacap**
- Faster I say! _Faster_!
- Turns out most of the slowdown is in `SyncFeed.Finish` that calls GC. This PR changes that so that it is only called on state sync finished.
- Also remove global lock in peer allocation.
- Also create a cached block tree builder that caches blocktree at db level.

#### What types of changes does your code introduce?

- [X] Optimization

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

#### Notes on testing

- Mainnet sync ok. Confirmed gc called once.

### CODE REVIEW

File: src/Nethermind/Nethermind.Core.Test/CachedBlockTreeBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            CachedDb cachedValue = new CachedDb(
+                MemDb.CopyFrom(builder.BlocksDb),
+                MemDb.CopyFrom(builder.MetadataDb),
+                MemDb.CopyFrom(builder.HeadersDb),
+                MemDb.CopyFrom(builder.BlockNumbersDb),
+                MemDb.CopyFrom(builder.BlockInfoDb)
+            );
+            _cachedDbs.TryAdd(key, cachedValue);
+
+            return builder.TestObject;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core.Test/CachedBlockTreeBuilder.cs:47]: Technically you should use `GetOrAdd` on `ConcurrentDictionary`

File: src/Nethermind/Nethermind.Synchronization/Peers/SyncPeerAllocation.cs [hunk 2]
```diff
@@ -29,14 +29,15 @@ public class SyncPeerAllocation
         public bool HasPeer => Current is not null;
 
         public SyncPeerAllocation(PeerInfo peerInfo, AllocationContexts contexts)
-            : this(new StaticStrategy(peerInfo), contexts)
+            : this(new StaticStrategy(peerInfo), contexts, null)
         {
         }
 
-        public SyncPeerAllocation(IPeerAllocationStrategy peerAllocationStrategy, AllocationContexts contexts)
+        public SyncPeerAllocation(IPeerAllocationStrategy peerAllocationStrategy, AllocationContexts contexts, Lock? allocationLock)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/Peers/SyncPeerAllocation.cs:36]: Make:
```suggestion
        public SyncPeerAllocation(IPeerAllocationStrategy peerAllocationStrategy, AllocationContexts contexts, Lock? allocationLock = null)
```

and don't pass `null` when not needed?

File: src/Nethermind/Nethermind.Synchronization/ParallelSync/SyncFeed.cs [hunk 3]
```diff
@@ -44,7 +44,6 @@ private void ChangeState(SyncFeedState newState)
         public virtual void Finish()
         {
             ChangeState(SyncFeedState.Finished);
-            GC.Collect(2, GCCollectionMode.Aggressive, true, true);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/ParallelSync/SyncFeed.cs:47]: I think this GC was also fired on Headers, Blocks and Receipts!

**Review by LukaszRozmej — COMMENTED** [review]
I think GC was not only for state?

File: src/Nethermind/Nethermind.Synchronization/ParallelSync/SyncFeed.cs (hunk 3 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/ParallelSync/SyncFeed.cs:47]: Thats fine, the state take most of the allocation.

Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/ParallelSync/SyncFeed.cs:47]: ok

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**